src/wallet/wallet.cpp

```diff
@@ -700,6 +700,16 @@ bool CWallet::IsSpent(const uint256& nullifier) const
     pair<TxNullifiers::const_iterator, TxNullifiers::const_iterator> range;
     range = mapTxSproutNullifiers.equal_range(nullifier);
 
+    for (TxNullifiers::const_iterator it = range.first; it != range.second; ++it) {
+        const uint256& wtxid = it->second;
+        std::map<uint256, CWalletTx>::const_iterator mit = mapWallet.find(wtxid);
+        if (mit != mapWallet.end() && mit->second.GetDepthInMainChain() >= 0) {
+            return true; // Spent
+        }
+    }
+
+    range = mapTxSaplingNullifiers.equal_range(nullifier);
```

Comment:
We need separate functions for checking Sprout and Sapling nullifiers, because they are in separate domains and aren't guaranteed to be collision-resistant (otherwise there is a possibility of a nullifier collision, however remote, between Sprout and Sapling causing the spend of one to prevent the spend of the other).#

Given how close to identical they would be, could implement this as a template (like was done for e.g. the anchor and tree handling code in `src/coins.h` IIRC).

Reply:
Good point.  No to templates!  Pushed commit for this.